internal/xds/translator/cluster.go

```diff
@@ -627,40 +630,61 @@ func buildXdsClusterCircuitBreaker(circuitBreaker *ir.CircuitBreaker) *clusterv3
 	return ecb
 }
 
-func buildXdsClusterLoadAssignment(clusterName string, destSettings []*ir.DestinationSetting) *endpointv3.ClusterLoadAssignment {
-	localities := make([]*endpointv3.LocalityLbEndpoints, 0, len(destSettings))
-	for i, ds := range destSettings {
-
-		var metadata *corev3.Metadata
+func getLbEndpointsTSMMetadata(ds *ir.DestinationSetting, dsIndex int, clusterName string) *corev3.Metadata {
+	if ds == nil || ds.TLS == nil {
+		return nil
+	}
 
-		if ds.TLS != nil {
-			metadata = &corev3.Metadata{
-				FilterMetadata: map[string]*structpb.Struct{
-					"envoy.transport_socket_match": {
-						Fields: map[string]*structpb.Value{
-							"name": structpb.NewStringValue(fmt.Sprintf("%s/tls/%d", clusterName, i)),
-						},
-					},
+	return &corev3.Metadata{
+		FilterMetadata: map[string]*structpb.Struct{
+			"envoy.transport_socket_match": {
+				Fields: map[string]*structpb.Value{
+					"name": structpb.NewStringValue(fmt.Sprintf("%s/tls/%d", clusterName, dsIndex)),
 				},
+			},
+		},
+	}
+}
+
+func buildXdsClusterLoadAssignment(clusterName string, destSettings []*ir.DestinationSetting) (*endpointv3.ClusterLoadAssignment, error) {
+	var localities []*endpointv3.LocalityLbEndpoints
+
+	hasZoneAwareRoutingEnabled := slices.ContainsFunc(destSettings, func(ds *ir.DestinationSetting) bool {
+		return ds.ZoneAwareRoutingEnabled
+	})
+
+	// If zone aware routing is enabled for a backendRefs we include endpoint zone info in localities.
+	// Note: The locality.LoadBalancingWeight field applies only when using locality-weighted LB config
+	// so in order to support traffic splitting we rely on weighted clusters defined at the route level
+	// if multiple backendRefs exist. This pushes part of the routing logic higher up the stack which can
+	// limit host selection controls during retries and session affinity.
+	// For more details see https://github.com/envoyproxy/gateway/issues/5307#issuecomment-2688767482
+
+	switch {
```

Comment:
we will not need these 2 cases, because you're logic is eliminating the need for weighted clusters altogether

Reply:
Zone Aware Routing still can't use the same logic because the `loadBalancingWeight` field at the locality level isn't respected. This PR adds endpoint zone information for clusters using locality weighted lb config. Because `loadBalancingWeight` _is_ respected, the same approach of using weighted clusters isn't needed. The tradeoff there is avoiding the drawbacks of weighted clusters for the cosmetic difference of weight normalization.